flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriteRequestDispatcherImpl.java

```diff
@@ -98,18 +120,19 @@ private ChannelStateCheckpointWriter buildWriter(CheckpointStartRequest request)
                 streamFactoryResolver.resolveCheckpointStorageLocation(
                         request.getCheckpointId(), request.getLocationReference()),
                 serializer,
-                () -> writers.remove(request.getCheckpointId()));
+                () -> this.writer = null);
```

Comment:
shouldn't you check here if this is the correct writer? Could it happen that the `this.writer` got reseted for a newer checkpoint but this callback happens later, reseting the field back to `null`? If this can not happen, it would be nice to at least have a `checkState` to make sure it's not happening?